File: src/shared/srs.ts
import { clozeIndexes } from './cloze';
import type { CardPhase, FlashCard, FlashNote, Rating, SrsDayStats } from './types';

/**
 * Anki-classic SM-2 scheduler. Pure functions only: every entry point takes
 * `now` (ms epoch) so tests inject time. No fuzz in v1 — if added later it
 * belongs at the end of answerCard, jittering review intervals by ±5%.
 */

export const LEARNING_STEPS_MIN = [1, 10] as const;
export const RELEARNING_STEPS_MIN = [10] as const;
export const GRADUATING_INTERVAL_DAYS = 1;
export const EASY_INTERVAL_DAYS = 4;
export const START_EASE = 2.5;
export const MIN_EASE = 1.3;
export const NEW_PER_DAY = 20;
export const MAX_INTERVAL_DAYS = 36500;
export const LEARN_AHEAD_MIN = 20;

const MIN_MS = 60_000;
const DAY_MS = 86_400_000;

export function newCard(noteId: string, deckId: string, variant: number, now: number): FlashCard {
  return {
    id: `${noteId}#${variant}`,
    noteId,
    deckId,
    variant,
    phase: 'new',
    stepIndex: 0,
    ease: START_EASE,
    intervalDays: 0,
    dueAt: now,
    lapses: 0,
    reps: 0,
    createdAt: now,
  };
}

function clampEase(ease: number): number {
  return Math.max(MIN_EASE, Math.round(ease * 100) / 100);
}

function clampInterval(days: number): number {
  return Math.min(MAX_INTERVAL_DAYS, days);
}

function graduate(card: FlashCard, intervalDays: number, now: number): FlashCard {
  return {
    ...card,
    phase: 'review',
    stepIndex: 0,
    intervalDays,
    dueAt: now + intervalDays * DAY_MS,
  };
}

export function answerCard(card: FlashCard, rating: Rating, now: number): FlashCard {
  const next = ((): FlashCard => {
    if (card.phase === 'new' || card.phase === 'learning') {
      const steps = LEARNING_STEPS_MIN;
      switch (rating) {
        case 'again':
          return { ...card, phase: 'learning', stepIndex: 0, dueAt: now + steps[0] * MIN_MS };
        case 'hard': {
          // Repeat the current step
          const step = steps[Math.min(card.stepIndex, steps.length - 1)];
          return { ...card, phase: 'learning', dueAt: now + step * MIN_MS };
        }
        case 'good': {
          const nextStep = card.phase === 'new' ? 1 : card.stepIndex + 1;
          if (nextStep >= steps.length) return graduate(card, GRADUATING_INTERVAL_DAYS, now);
          return { ...card, phase: 'learning', stepIndex: nextStep, dueAt: now + steps[nextStep] * MIN_MS };
        }
        case 'easy':
          return graduate(card, EASY_INTERVAL_DAYS, now);
      }
    }

    if (card.phase === 'review') {
      const i = card.intervalDays;
      switch (rating) {
        case 'again':
          // Lapse: ease penalty, relearn steps; post-relearn interval is 1 day
          return {
            ...card,
            phase: 'relearning',
            stepIndex: 0,
            ease: clampEase(card.ease - 0.2),
            lapses: card.lapses + 1,
            intervalDays: 1,
            dueAt: now + RELEARNING_STEPS_MIN[0] * MIN_MS,
          };
        case 'hard': {
          const interval = clampInterval(Math.max(i + 1, Math.round(i * 1.2)));
          return {
            ...card,
            ease: clampEase(card.ease - 0.15),
            intervalDays: interval,
            dueAt: now + interval * DAY_MS,
          };
        }
        case 'good': {
          const interval = clampInterval(Math.max(i + 1, Math.round(i * card.ease)));
          return { ...card, intervalDays: interval, dueAt: now + interval * DAY_MS };
        }
        case 'easy': {
          const interval = clampInterval(Math.max(i + 1, Math.round(i * card.ease * 1.3)));
          return {
            ...card,
            ease: clampEase(card.ease + 0.15),
            intervalDays: interval,
            dueAt: now + interval * DAY_MS,
          };
        }
      }
    }

    // relearning
    const steps = RELEARNING_STEPS_MIN;
    switch (rating) {
      case 'again':
        // No further ease penalty on relearn misses (Anki behavior)
        return { ...card, stepIndex: 0, dueAt: now + steps[0] * MIN_MS };
      case 'hard': {
        const step = steps[Math.min(card.stepIndex, steps.length - 1)];
        return { ...card, dueAt: now + step * MIN_MS };
      }
      case 'good': {
        const nextStep = card.stepIndex + 1;
        if (nextStep >= steps.length) return graduate(card, card.intervalDays, now);
        return { ...card, stepIndex: nextStep, dueAt: now + steps[nextStep] * MIN_MS };
      }
      case 'easy':
        return graduate(card, card.intervalDays, now);
    }
  })();

  return { ...next, reps: card.reps + 1 };
}

/** Human label for a duration, matching Anki's answer-button previews */
export function formatInterval(ms: number): string {
  const mins = Math.round(ms / MIN_MS);
  if (mins < 60) return `${Math.max(1, mins)}m`;
  const days = ms / DAY_MS;
  if (days < 1) return `${Math.round(days * 24)}h`;
  if (days < 30) return `${Math.round(days)}d`;
  if (days < 365) return `${(days / 30.44).toFixed(1).replace(/\.0$/, '')}mo`;
  return `${(days / 365.25).toFixed(1).replace(/\.0$/, '')}yr`;
}

export function previewIntervals(card: FlashCard, now: number): Record<Rating, string> {
  const out = {} as Record<Rating, string>;
  for (const rating of ['again', 'hard', 'good', 'easy'] as const) {
    out[rating] = formatInterval(answerCard(card, rating, now).dueAt - now);
  }
  return out;
}

/** Variant numbers a note should have cards for */
export function cardsForNote(note: FlashNote): number[] {
  if (note.type === 'cloze') return clozeIndexes(note.front);
  return note.reversed ? [0, 1] : [0];
}

/**
 * Reconcile a note's cards after create/edit: surviving variants keep their
 * scheduling state, new variants start fresh, removed variants are dropped.
 */
export function reconcileCards(note: FlashNote, existing: FlashCard[], now: number): FlashCard[] {
  const byVariant = new Map(existing.map((c) => [c.variant, c]));
  return cardsForNote(note).map(
    (variant) =>
      byVariant.get(variant) ?? newCard(note.id, note.deckId, variant, now),
  );
}

/** Local end of day (exclusive): review cards are "due today" until local midnight */
export function endOfLocalDay(now: number): number {
  const d = new Date(now);
  d.setHours(24, 0, 0, 0);
  return d.getTime();
}

/**
 * Ordered study queue for one deck:
 * 1. learning/relearning cards already due,
 * 2. new cards (oldest first) within the remaining daily allowance,
 * 3. review cards due today,
 * 4. if all else is empty, learning cards due within LEARN_AHEAD_MIN.
 */
export function buildQueue(
  cards: FlashCard[],
  deckId: string,
  now: number,
  newIntroducedToday: number,
): FlashCard[] {
  const deck = cards.filter((c) => c.deckId === deckId);
  const byDue = (a: FlashCard, b: FlashCard) => a.dueAt - b.dueAt;

  const learning = deck
    .filter((c) => (c.phase === 'learning' || c.phase === 'relearning') && c.dueAt <= now)
    .sort(byDue);
  const fresh = deck
    .filter((c) => c.phase === 'new')
    .sort((a, b) => a.createdAt - b.createdAt)
    .slice(0, Math.max(0, NEW_PER_DAY - newIntroducedToday));
  const review = deck
    .filter((c) => c.phase === 'review' && c.dueAt <= endOfLocalDay(now))
    .sort(byDue);

  const queue = [...learning, ...fresh, ...review];
  if (queue.length > 0) return queue;

  return deck
    .filter(
      (c) =>
        (c.phase === 'learning' || c.phase === 'relearning') &&
        c.dueAt <= now + LEARN_AHEAD_MIN * MIN_MS,
    )
    .sort(byDue);
}

export interface DeckDueCounts {
  newCount: number;
  learningCount: number;
  reviewCount: number;
}

/** Due counts per deck — shared by deck list, dashboard card, and popup tab */
export function dueCounts(
  cards: FlashCard[],
  now: number,
  newIntroducedByDeck: Record<string, number>,
): Record<string, DeckDueCounts> {
  const endOfDay = endOfLocalDay(now);
  const out: Record<string, DeckDueCounts> = {};
  for (const card of cards) {
    const counts = (out[card.deckId] ??= { newCount: 0, learningCount: 0, reviewCount: 0 });
    if (card.phase === 'new') counts.newCount += 1;
    else if ((card.phase === 'learning' || card.phase === 'relearning') && card.dueAt <= now)
      counts.learningCount += 1;
    else if (card.phase === 'review' && card.dueAt <= endOfDay) counts.reviewCount += 1;
  }
  for (const [deckId, counts] of Object.entries(out)) {
    const allowance = Math.max(0, NEW_PER_DAY - (newIntroducedByDeck[deckId] ?? 0));
    counts.newCount = Math.min(counts.newCount, allowance);
  }
  return out;
}

/** Total cards due across all decks (dashboard headline / popup badge) */
export function totalDue(counts: Record<string, DeckDueCounts>): number {
  return Object.values(counts).reduce(
    (sum, c) => sum + c.newCount + c.learningCount + c.reviewCount,
    0,
  );
}

/** newIntroduced-by-deck for today, from the srsDaily aggregate */
export function newIntroducedToday(
  srsDaily: Record<string, SrsDayStats>,
  todayKey: string,
): Record<string, number> {
  return srsDaily[todayKey]?.newIntroduced ?? {};
}

/** True when answering this card should award XP (see gamification design) */
export function isRewardableAnswer(prevPhase: CardPhase, next: FlashCard): boolean {
  if (prevPhase === 'review') return true;
  return (prevPhase === 'new' || prevPhase === 'learning') && next.phase === 'review';
}

File: ios/ADHDReaderCore/Sources/ADHDReaderCore/SRS.swift
import Foundation

/// Anki-classic SM-2 scheduler, ported 1:1 from `src/shared/srs.ts`. Pure:
/// every entry point takes `now` (ms epoch) so tests inject time. SRSTests
/// mirrors the Vitest suite.
public enum SRS {
    public static let learningStepsMin = [1, 10]
    public static let relearningStepsMin = [10]
    public static let graduatingIntervalDays = 1
    public static let easyIntervalDays = 4
    public static let startEase = 2.5
    public static let minEase = 1.3
    public static let newPerDay = 20
    public static let maxIntervalDays = 36_500
    public static let learnAheadMin = 20

    static let minMs = 60_000
    static let dayMs = 86_400_000

    public static func newCard(noteId: String, deckId: String, variant: Int, now: Millis) -> FlashCard {
        FlashCard(
            id: "\(noteId)#\(variant)", noteId: noteId, deckId: deckId, variant: variant,
            phase: .new, stepIndex: 0, ease: startEase, intervalDays: 0, dueAt: now,
            lapses: 0, reps: 0, createdAt: now
        )
    }

    private static func clampEase(_ ease: Double) -> Double {
        max(minEase, (ease * 100).rounded() / 100)
    }

    private static func clampInterval(_ days: Int) -> Int { min(maxIntervalDays, days) }

    private static func graduate(_ card: FlashCard, _ intervalDays: Int, _ now: Millis) -> FlashCard {
        var c = card
        c.phase = .review
        c.stepIndex = 0
        c.intervalDays = intervalDays
        c.dueAt = now + intervalDays * dayMs
        return c
    }

    public static func answerCard(_ card: FlashCard, _ rating: Rating, now: Millis) -> FlashCard {
        var next = computeNext(card, rating, now)
        next.reps = card.reps + 1
        return next
    }

    private static func computeNext(_ card: FlashCard, _ rating: Rating, _ now: Millis) -> FlashCard {
        var c = card
        if card.phase == .new || card.phase == .learning {
            let steps = learningStepsMin
            switch rating {
            case .again:
                c.phase = .learning; c.stepIndex = 0; c.dueAt = now + steps[0] * minMs
                return c
            case .hard:
                let step = steps[min(card.stepIndex, steps.count - 1)]
                c.phase = .learning; c.dueAt = now + step * minMs
                return c
            case .good:
                let nextStep = card.phase == .new ? 1 : card.stepIndex + 1
                if nextStep >= steps.count { return graduate(card, graduatingIntervalDays, now) }
                c.phase = .learning; c.stepIndex = nextStep; c.dueAt = now + steps[nextStep] * minMs
                return c
            case .easy:
                return graduate(card, easyIntervalDays, now)
            }
        }

        if card.phase == .review {
            let i = card.intervalDays
            switch rating {
            case .again:
                c.phase = .relearning; c.stepIndex = 0
                c.ease = clampEase(card.ease - 0.2); c.lapses = card.lapses + 1
                c.intervalDays = 1; c.dueAt = now + relearningStepsMin[0] * minMs
                return c
            case .hard:
                let interval = clampInterval(max(i + 1, Int((Double(i) * 1.2).rounded())))
                c.ease = clampEase(card.ease - 0.15)
                c.intervalDays = interval; c.dueAt = now + interval * dayMs
                return c
            case .good:
                let interval = clampInterval(max(i + 1, Int((Double(i) * card.ease).rounded())))
                c.intervalDays = interval; c.dueAt = now + interval * dayMs
                return c
            case .easy:
                let interval = clampInterval(max(i + 1, Int((Double(i) * card.ease * 1.3).rounded())))
                c.ease = clampEase(card.ease + 0.15)
                c.intervalDays = interval; c.dueAt = now + interval * dayMs
                return c
            }
        }

        // relearning
        let steps = relearningStepsMin
        switch rating {
        case .again:
            c.stepIndex = 0; c.dueAt = now + steps[0] * minMs
            return c
        case .hard:
            let step = steps[min(card.stepIndex, steps.count - 1)]
            c.dueAt = now + step * minMs
            return c
        case .good:
            let nextStep = card.stepIndex + 1
            if nextStep >= steps.count { return graduate(card, card.intervalDays, now) }
            c.stepIndex = nextStep; c.dueAt = now + steps[nextStep] * minMs
            return c
        case .easy:
            return graduate(card, card.intervalDays, now)
        }
    }

    /// Human label for a duration, matching Anki's answer-button previews.
    public static func formatInterval(_ ms: Int) -> String {
        let mins = Int((Double(ms) / Double(minMs)).rounded())
        if mins < 60 { return "\(max(1, mins))m" }
        let days = Double(ms) / Double(dayMs)
        if days < 1 { return "\(Int((days * 24).rounded()))h" }
        if days < 30 { return "\(Int(days.rounded()))d" }
        if days < 365 { return "\(trimmed(days / 30.44))mo" }
        return "\(trimmed(days / 365.25))yr"
    }

    /// One decimal place, trailing ".0" stripped (matches JS `.toFixed(1).replace`).
    private static func trimmed(_ value: Double) -> String {
        let s = String(format: "%.1f", value)
        return s.hasSuffix(".0") ? String(s.dropLast(2)) : s
    }

    public static func previewIntervals(_ card: FlashCard, now: Millis) -> [Rating: String] {
        var out: [Rating: String] = [:]
        for rating in Rating.allCases {
            out[rating] = formatInterval(answerCard(card, rating, now: now).dueAt - now)
        }
        return out
    }

    /// Local end of day (exclusive): review cards are "due today" until local midnight.
    public static func endOfLocalDay(_ now: Millis) -> Millis {
        let date = Date(timeIntervalSince1970: Double(now) / 1000)
        let cal = Calendar.current
        let start = cal.startOfDay(for: date)
        let next = cal.date(byAdding: .day, value: 1, to: start)!
        return Int((next.timeIntervalSince1970 * 1000).rounded())
    }

    /// Ordered study queue for one deck (learning → new within allowance →
    /// review due today → learn-ahead fallback).
    public static func buildQueue(
        _ cards: [FlashCard], deckId: String, now: Millis, newIntroducedToday: Int
    ) -> [FlashCard] {
        let deck = cards.filter { $0.deckId == deckId }
        let byDue: (FlashCard, FlashCard) -> Bool = { $0.dueAt < $1.dueAt }

        let learning = deck
            .filter { ($0.phase == .learning || $0.phase == .relearning) && $0.dueAt <= now }
            .sorted(by: byDue)
        let fresh = deck
            .filter { $0.phase == .new }
            .sorted { $0.createdAt < $1.createdAt }
            .prefix(max(0, newPerDay - newIntroducedToday))
        let endOfDay = endOfLocalDay(now)
        let review = deck
            .filter { $0.phase == .review && $0.dueAt <= endOfDay }
            .sorted(by: byDue)

        let queue = learning + Array(fresh) + review
        if !queue.isEmpty { return queue }

        return deck
            .filter {
                ($0.phase == .learning || $0.phase == .relearning)
                    && $0.dueAt <= now + learnAheadMin * minMs
            }
            .sorted(by: byDue)
    }

    public struct DeckDueCounts: Equatable, Sendable {
        public var newCount: Int
        public var learningCount: Int
        public var reviewCount: Int
        public init(newCount: Int = 0, learningCount: Int = 0, reviewCount: Int = 0) {
            self.newCount = newCount; self.learningCount = learningCount; self.reviewCount = reviewCount
        }
    }

    /// Due counts per deck — shared by deck list, dashboard card, and popup tab.
    public static func dueCounts(
        _ cards: [FlashCard], now: Millis, newIntroducedByDeck: [String: Int]
    ) -> [String: DeckDueCounts] {
        let endOfDay = endOfLocalDay(now)
        var out: [String: DeckDueCounts] = [:]
        for card in cards {
            var counts = out[card.deckId] ?? DeckDueCounts()
            if card.phase == .new {
                counts.newCount += 1
            } else if (card.phase == .learning || card.phase == .relearning) && card.dueAt <= now {
                counts.learningCount += 1
            } else if card.phase == .review && card.dueAt <= endOfDay {
                counts.reviewCount += 1
            }
            out[card.deckId] = counts
        }
        for (deckId, var counts) in out {
            let allowance = max(0, newPerDay - (newIntroducedByDeck[deckId] ?? 0))
            counts.newCount = min(counts.newCount, allowance)
            out[deckId] = counts
        }
        return out
    }

    /// Total cards due across all decks.
    public static func totalDue(_ counts: [String: DeckDueCounts]) -> Int {
        counts.values.reduce(0) { $0 + $1.newCount + $1.learningCount + $1.reviewCount }
    }

    /// True when answering this card should award XP (see gamification design).
    public static func isRewardableAnswer(prevPhase: CardPhase, next: FlashCard) -> Bool {
        if prevPhase == .review { return true }
        return (prevPhase == .new || prevPhase == .learning) && next.phase == .review
    }
}
